File: packages/analytics/labeling.py
"""Note-topic labeling using embedding similarity."""

from dataclasses import dataclass
from typing import Any

import numpy as np
from psycopg import AsyncConnection

from packages.analytics.taxonomy import Taxonomy
from packages.common.config import Settings, get_settings
from packages.common.database import get_connection
from packages.indexer.embeddings import EmbeddingProvider, get_embedding_provider
from packages.indexer.qdrant import QdrantRepository, get_qdrant_repository


@dataclass
class TopicAssignment:
    """A topic assignment for a note."""

    note_id: int
    topic_id: int
    topic_path: str
    confidence: float
    method: str = "embedding"


@dataclass
class LabelingStats:
    """Statistics from a labeling operation."""

    notes_processed: int = 0
    assignments_created: int = 0
    topics_matched: int = 0


def cosine_similarity(vec1: list[float], vec2: list[float]) -> float:
    """Compute cosine similarity between two vectors."""
    a = np.array(vec1)
    b = np.array(vec2)
    return float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)))


class TopicLabeler:
    """Service for labeling notes with topics."""

    def __init__(
        self,
        settings: Settings | None = None,
        embedding_provider: EmbeddingProvider | None = None,
        qdrant_repository: QdrantRepository | None = None,
    ) -> None:
        """Initialize labeler.

        Args:
            settings: Application settings.
            embedding_provider: Embedding provider.
            qdrant_repository: Qdrant repository.
        """
        self.settings = settings or get_settings()
        self._embedding_provider = embedding_provider
        self._qdrant_repository = qdrant_repository

    async def get_embedding_provider(self) -> EmbeddingProvider:
        """Get or create embedding provider."""
        if self._embedding_provider is None:
            self._embedding_provider = get_embedding_provider(self.settings)
        return self._embedding_provider

    async def get_qdrant_repository(self) -> QdrantRepository:
        """Get or create Qdrant repository."""
        if self._qdrant_repository is None:
            self._qdrant_repository = await get_qdrant_repository(self.settings)
        return self._qdrant_repository

    async def embed_topics(
        self,
        taxonomy: Taxonomy,
    ) -> dict[str, list[float]]:
        """Embed all topic descriptions.

        Args:
            taxonomy: Taxonomy with topics to embed.

        Returns:
            Dictionary mapping topic path to embedding vector.
        """
        provider = await self.get_embedding_provider()

        # Build texts to embed (use description or label)
        topic_texts: list[tuple[str, str]] = []
        for topic in taxonomy.all_topics():
            text = topic.description or topic.label
            topic_texts.append((topic.path, text))

        if not topic_texts:
            return {}

        # Embed all texts
        texts = [t[1] for t in topic_texts]
        vectors = await provider.embed(texts)

        return {topic_texts[i][0]: vectors[i] for i in range(len(topic_texts))}

    async def label_notes(
        self,
        taxonomy: Taxonomy,
        min_confidence: float = 0.3,
        max_topics_per_note: int = 3,
        batch_size: int = 100,
    ) -> LabelingStats:
        """Label all notes with matching topics.

        Args:
            taxonomy: Taxonomy to use for labeling.
            min_confidence: Minimum similarity threshold.
            max_topics_per_note: Maximum topics to assign per note.
            batch_size: Batch size for processing.

        Returns:
            Labeling statistics.
        """
        stats = LabelingStats()

        # Embed topics
        topic_embeddings = await self.embed_topics(taxonomy)
        if not topic_embeddings:
            return stats

        stats.topics_matched = len(topic_embeddings)

        # Get all note embeddings from database
        async with get_connection(self.settings) as conn:
            # Count notes
            result = await conn.execute(
                "SELECT COUNT(*) as count FROM notes WHERE deleted_at IS NULL"
            )
            row = await result.fetchone()
            total_notes = row["count"] if row else 0

            # Process in batches
            offset = 0
            while offset < total_notes:
                assignments = await self._label_batch(
                    conn,
                    taxonomy,
                    topic_embeddings,
                    offset,
                    batch_size,
                    min_confidence,
                    max_topics_per_note,
                )

                stats.notes_processed += batch_size
                stats.assignments_created += len(assignments)

                # Store assignments
                if assignments:
                    await self._store_assignments(conn, assignments)

                offset += batch_size

            await conn.commit()

        return stats

    async def _label_batch(
        self,
        conn: AsyncConnection[dict[str, Any]],
        taxonomy: Taxonomy,
        topic_embeddings: dict[str, list[float]],
        offset: int,
        limit: int,
        min_confidence: float,
        max_topics_per_note: int,
    ) -> list[TopicAssignment]:
        """Label a batch of notes."""
        provider = await self.get_embedding_provider()

        # Fetch notes
        result = await conn.execute(
            """
            SELECT note_id, normalized_text
            FROM notes
            WHERE deleted_at IS NULL
            ORDER BY note_id
            LIMIT %(limit)s OFFSET %(offset)s
            """,
            {"limit": limit, "offset": offset},
        )

        notes: list[tuple[int, str]] = []
        async for row in result:
            notes.append((row["note_id"], row["normalized_text"]))

        if not notes:
            return []

        # Embed note texts
        note_texts = [n[1] for n in notes]
        note_vectors = await provider.embed(note_texts)

        # Find best topics for each note
        assignments: list[TopicAssignment] = []
        topic_paths = list(topic_embeddings.keys())
        topic_vectors = [topic_embeddings[p] for p in topic_paths]

        for i, (note_id, _) in enumerate(notes):
            note_vec = note_vectors[i]

            # Compute similarities to all topics
            similarities: list[tuple[str, float]] = []
            for j, topic_path in enumerate(topic_paths):
                sim = cosine_similarity(note_vec, topic_vectors[j])
                if sim >= min_confidence:
                    similarities.append((topic_path, sim))

            # Sort by similarity and take top N
            similarities.sort(key=lambda x: x[1], reverse=True)
            top_matches = similarities[:max_topics_per_note]

            for topic_path, confidence in top_matches:
                topic = taxonomy.get(topic_path)
                if topic and topic.topic_id:
                    assignments.append(
                        TopicAssignment(
                            note_id=note_id,
                            topic_id=topic.topic_id,
                            topic_path=topic_path,
                            confidence=confidence,
                        )
                    )

        return assignments

    async def _store_assignments(
        self,
        conn: AsyncConnection[dict[str, Any]],
        assignments: list[TopicAssignment],
    ) -> None:
        """Store topic assignments in database."""
        for assignment in assignments:
            await conn.execute(
                """
                INSERT INTO note_topics (note_id, topic_id, confidence, method)
                VALUES (%(note_id)s, %(topic_id)s, %(confidence)s, %(method)s)
                ON CONFLICT (note_id, topic_id) DO UPDATE SET
                    confidence = EXCLUDED.confidence,
                    method = EXCLUDED.method
                """,
                {
                    "note_id": assignment.note_id,
                    "topic_id": assignment.topic_id,
                    "confidence": assignment.confidence,
                    "method": assignment.method,
                },
            )

    async def label_single_note(
        self,
        note_id: int,
        taxonomy: Taxonomy,
        topic_embeddings: dict[str, list[float]] | None = None,
        min_confidence: float = 0.3,
        max_topics: int = 3,
    ) -> list[TopicAssignment]:
        """Label a single note with topics.

        Args:
            note_id: Note ID to label.
            taxonomy: Taxonomy to use.
            topic_embeddings: Pre-computed topic embeddings (optional).
            min_confidence: Minimum similarity threshold.
            max_topics: Maximum topics to assign.

        Returns:
            List of topic assignments.
        """
        provider = await self.get_embedding_provider()

        # Get topic embeddings if not provided
        if topic_embeddings is None:
            topic_embeddings = await self.embed_topics(taxonomy)

        if not topic_embeddings:
            return []

        # Get note text
        async with get_connection(self.settings) as conn:
            result = await conn.execute(
                "SELECT normalized_text FROM notes WHERE note_id = %(note_id)s",
                {"note_id": note_id},
            )
            row = await result.fetchone()
            if not row:
                return []

            note_text = row["normalized_text"]

        # Embed note
        note_vector = await provider.embed_single(note_text)

        # Find matching topics
        assignments: list[TopicAssignment] = []
        for topic_path, topic_vec in topic_embeddings.items():
            sim = cosine_similarity(note_vector, topic_vec)
            if sim >= min_confidence:
                topic = taxonomy.get(topic_path)
                if topic and topic.topic_id:
                    assignments.append(
                        TopicAssignment(
                            note_id=note_id,
                            topic_id=topic.topic_id,
                            topic_path=topic_path,
                            confidence=sim,
                        )
                    )

        # Sort and limit
        assignments.sort(key=lambda x: x.confidence, reverse=True)
        return assignments[:max_topics]


async def label_all_notes(
    taxonomy: Taxonomy,
    settings: Settings | None = None,
    min_confidence: float = 0.3,
) -> LabelingStats:
    """Convenience function to label all notes.

    Args:
        taxonomy: Taxonomy to use.
        settings: Application settings.
        min_confidence: Minimum confidence threshold.

    Returns:
        Labeling statistics.
    """
    labeler = TopicLabeler(settings)
    return await labeler.label_notes(taxonomy, min_confidence=min_confidence)

File: crates/analytics/src/labeling.rs
use std::collections::HashMap;

use serde::Serialize;

use crate::AnalyticsError;

/// Method used to assign a topic to a note.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum LabelingMethod {
    Embedding,
}

impl LabelingMethod {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Embedding => "embedding",
        }
    }
}

/// A topic assignment for a note.
#[derive(Debug, Clone, Serialize)]
pub struct TopicAssignment {
    pub note_id: i64,
    pub topic_id: i64,
    pub topic_path: String,
    pub confidence: f64,
    pub method: LabelingMethod,
}

/// Statistics from a labeling operation.
#[derive(Debug, Clone, Default, Serialize)]
pub struct LabelingStats {
    pub notes_processed: usize,
    pub assignments_created: usize,
    pub topics_matched: usize,
}

/// Cosine similarity between two equal-length f32 vectors.
pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    let mut dot = 0.0_f32;
    let mut norm_a = 0.0_f32;
    let mut norm_b = 0.0_f32;
    for (ai, bi) in a.iter().zip(b.iter()) {
        dot += ai * bi;
        norm_a += ai * ai;
        norm_b += bi * bi;
    }
    let denom = norm_a.sqrt() * norm_b.sqrt();
    if denom == 0.0 { 0.0 } else { dot / denom }
}

/// Rank topics by cosine similarity to a note embedding.
/// Returns assignments above `min_confidence`, limited to `max_topics`,
/// sorted by confidence descending.
#[allow(dead_code)]
pub(crate) fn rank_topics_for_note(
    note_id: i64,
    note_embedding: &[f32],
    topic_embeddings: &HashMap<String, (i64, Vec<f32>)>,
    min_confidence: f32,
    max_topics: usize,
) -> Vec<TopicAssignment> {
    let mut assignments: Vec<TopicAssignment> = topic_embeddings
        .iter()
        .filter_map(|(path, (topic_id, emb))| {
            let sim = cosine_similarity(note_embedding, emb);
            if sim >= min_confidence {
                Some(TopicAssignment {
                    note_id,
                    topic_id: *topic_id,
                    topic_path: path.clone(),
                    confidence: f64::from(sim),
                    method: LabelingMethod::Embedding,
                })
            } else {
                None
            }
        })
        .collect();
    assignments.sort_by(|a, b| {
        b.confidence
            .partial_cmp(&a.confidence)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    assignments.truncate(max_topics);
    assignments
}

/// Topic labeler. Generic over embedding provider.
pub struct TopicLabeler<E: indexer::embeddings::EmbeddingProvider> {
    pub embedding: E,
    pub db: sqlx::PgPool,
}

impl<E: indexer::embeddings::EmbeddingProvider> TopicLabeler<E> {
    pub fn new(embedding: E, db: sqlx::PgPool) -> Self {
        Self { embedding, db }
    }

    /// Embed all topic descriptions/labels. Returns path -> embedding vector.
    pub async fn embed_topics(
        &self,
        taxonomy: &super::taxonomy::Taxonomy,
    ) -> Result<HashMap<String, Vec<f32>>, AnalyticsError> {
        let topic_paths: Vec<String> = taxonomy.topics.keys().cloned().collect();
        let topic_texts: Vec<String> = topic_paths
            .iter()
            .map(|path| {
                let topic = &taxonomy.topics[path];
                match &topic.description {
                    Some(desc) => format!("{}: {desc}", topic.label),
                    None => topic.label.clone(),
                }
            })
            .collect();

        let embeddings = self.embedding.embed(&topic_texts).await?;

        let result: HashMap<String, Vec<f32>> = topic_paths.into_iter().zip(embeddings).collect();

        Ok(result)
    }

    fn build_topic_embedding_index(
        taxonomy: &super::taxonomy::Taxonomy,
        topic_embeddings: &HashMap<String, Vec<f32>>,
    ) -> HashMap<String, (i64, Vec<f32>)> {
        topic_embeddings
            .iter()
            .filter_map(|(topic_path, embedding)| {
                taxonomy.topics.get(topic_path).and_then(|topic| {
                    topic
                        .topic_id
                        .map(|topic_id| (topic_path.clone(), (topic_id, embedding.clone())))
                })
            })
            .collect()
    }

    /// Label all notes in database with matching topics.
    pub async fn label_notes(
        &self,
        taxonomy: &super::taxonomy::Taxonomy,
        min_confidence: f32,
        max_topics_per_note: usize,
        batch_size: usize,
    ) -> Result<LabelingStats, AnalyticsError> {
        let topic_embeddings = self.embed_topics(taxonomy).await?;
        let topic_embedding_index = Self::build_topic_embedding_index(taxonomy, &topic_embeddings);

        let mut stats = LabelingStats::default();
        let mut matched_topics = std::collections::HashSet::new();
        let mut offset: i64 = 0;

        loop {
            let notes: Vec<(i64, String)> = sqlx::query_as(
                "SELECT note_id, normalized_text FROM notes \
                 WHERE deleted_at IS NULL \
                 ORDER BY note_id LIMIT $1 OFFSET $2",
            )
            .bind(batch_size as i64)
            .bind(offset)
            .fetch_all(&self.db)
            .await?;

            if notes.is_empty() {
                break;
            }

            let texts: Vec<String> = notes.iter().map(|(_, text)| text.clone()).collect();
            let note_embeddings = self.embedding.embed(&texts).await?;

            for ((note_id, _), note_emb) in notes.iter().zip(note_embeddings.iter()) {
                let assignments = rank_topics_for_note(
                    *note_id,
                    note_emb,
                    &topic_embedding_index,
                    min_confidence,
                    max_topics_per_note,
                );

                for assignment in &assignments {
                    sqlx::query(
                        "INSERT INTO note_topics (note_id, topic_id, confidence, method) \
                         VALUES ($1, $2, $3, $4) \
                         ON CONFLICT (note_id, topic_id) DO UPDATE \
                         SET confidence = $3, method = $4",
                    )
                    .bind(assignment.note_id)
                    .bind(assignment.topic_id as i32)
                    .bind(assignment.confidence as f32)
                    .bind(assignment.method.as_str())
                    .execute(&self.db)
                    .await?;

                    matched_topics.insert(assignment.topic_path.clone());
                }

                stats.assignments_created += assignments.len();
                stats.notes_processed += 1;
            }

            offset += notes.len() as i64;
        }

        stats.topics_matched = matched_topics.len();
        Ok(stats)
    }

    /// Label a single note.
    pub async fn label_single_note(
        &self,
        note_id: i64,
        taxonomy: &super::taxonomy::Taxonomy,
        topic_embeddings: Option<&HashMap<String, Vec<f32>>>,
        min_confidence: f32,
        max_topics: usize,
    ) -> Result<Vec<TopicAssignment>, AnalyticsError> {
        // Get note text
        let (text,): (String,) =
            sqlx::query_as("SELECT normalized_text FROM notes WHERE note_id = $1")
                .bind(note_id)
                .fetch_one(&self.db)
                .await?;

        // Embed the note
        let note_embeddings = self.embedding.embed(&[text]).await?;
        let note_emb = &note_embeddings[0];

        // Get or compute topic embeddings
        let owned_topic_embeddings;
        let topic_embeddings = match topic_embeddings {
            Some(existing_topic_embeddings) => existing_topic_embeddings,
            None => {
                owned_topic_embeddings = self.embed_topics(taxonomy).await?;
                &owned_topic_embeddings
            }
        };

        let topic_embedding_index = Self::build_topic_embedding_index(taxonomy, topic_embeddings);

        let assignments = rank_topics_for_note(
            note_id,
            note_emb,
            &topic_embedding_index,
            min_confidence,
            max_topics,
        );

        Ok(assignments)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cosine_similarity_identical_vectors() {
        let v = vec![1.0_f32, 2.0, 3.0];
        let sim = cosine_similarity(&v, &v);
        assert!((sim - 1.0).abs() < 1e-5, "expected ~1.0, got {sim}");
    }

    #[test]
    fn cosine_similarity_orthogonal_vectors() {
        let a = vec![1.0_f32, 0.0, 0.0];
        let b = vec![0.0_f32, 1.0, 0.0];
        let sim = cosine_similarity(&a, &b);
        assert!(sim.abs() < 1e-5, "expected ~0.0, got {sim}");
    }

    #[test]
    fn cosine_similarity_opposite_vectors() {
        let a = vec![1.0_f32, 0.0];
        let b = vec![-1.0_f32, 0.0];
        let sim = cosine_similarity(&a, &b);
        assert!((sim + 1.0).abs() < 1e-5, "expected ~-1.0, got {sim}");
    }

    #[test]
    fn cosine_similarity_known_value() {
        // cos(45 degrees) = sqrt(2)/2 ~ 0.7071
        let a = vec![1.0_f32, 0.0];
        let b = vec![1.0_f32, 1.0];
        let sim = cosine_similarity(&a, &b);
        let expected = std::f32::consts::FRAC_1_SQRT_2;
        assert!(
            (sim - expected).abs() < 1e-5,
            "expected ~{expected}, got {sim}"
        );
    }

    #[test]
    fn cosine_similarity_unit_vectors() {
        let a = vec![0.6_f32, 0.8]; // unit vector
        let sim = cosine_similarity(&a, &a);
        assert!((sim - 1.0).abs() < 1e-5, "expected ~1.0, got {sim}");
    }

    // --- rank_topics_for_note ---

    fn make_topic_embeddings(
        entries: &[(&str, i64, Vec<f32>)],
    ) -> HashMap<String, (i64, Vec<f32>)> {
        entries
            .iter()
            .map(|(path, id, emb)| (path.to_string(), (*id, emb.clone())))
            .collect()
    }

    #[test]
    fn rank_topics_skips_below_min_confidence() {
        // Note embedding is unit vector along x-axis
        let note_emb = vec![1.0_f32, 0.0, 0.0];
        // Topic A: nearly aligned (high similarity ~0.98)
        // Topic B: mostly orthogonal (low similarity ~0.17)
        let topics = make_topic_embeddings(&[
            ("topicA", 1, vec![0.98, 0.2, 0.0]),
            ("topicB", 2, vec![0.1, 0.98, 0.17]),
        ]);
        let result = rank_topics_for_note(42, &note_emb, &topics, 0.8, 10);
        // Only topicA should pass the 0.8 threshold
        assert_eq!(result.len(), 1, "should skip topics below min_confidence");
        assert_eq!(result[0].topic_path, "topicA");
        assert!(result[0].confidence >= 0.8);
    }

    #[test]
    fn rank_topics_respects_max_topics_limit() {
        let note_emb = vec![1.0_f32, 0.0, 0.0];
        // All topics have high similarity
        let topics = make_topic_embeddings(&[
            ("a", 1, vec![1.0, 0.0, 0.0]),
            ("b", 2, vec![0.99, 0.14, 0.0]),
            ("c", 3, vec![0.98, 0.2, 0.0]),
            ("d", 4, vec![0.97, 0.24, 0.0]),
        ]);
        let result = rank_topics_for_note(42, &note_emb, &topics, 0.0, 2);
        assert_eq!(result.len(), 2, "should limit to max_topics=2");
    }

    #[test]
    fn rank_topics_sorted_by_confidence_descending() {
        let note_emb = vec![1.0_f32, 0.0, 0.0];
        let topics = make_topic_embeddings(&[
            ("low", 1, vec![0.7, 0.7, 0.0]),
            ("high", 2, vec![1.0, 0.0, 0.0]),
            ("mid", 3, vec![0.9, 0.4, 0.0]),
        ]);
        let result = rank_topics_for_note(42, &note_emb, &topics, 0.0, 10);
        assert!(result.len() >= 2);
        for w in result.windows(2) {
            assert!(
                w[0].confidence >= w[1].confidence,
                "expected descending order: {} >= {}",
                w[0].confidence,
                w[1].confidence
            );
        }
    }

    #[test]
    fn rank_topics_empty_topic_embeddings() {
        let note_emb = vec![1.0_f32, 0.0];
        let topics = HashMap::new();
        let result = rank_topics_for_note(42, &note_emb, &topics, 0.0, 10);
        assert!(result.is_empty());
    }

    #[test]
    fn rank_topics_sets_method_to_embedding() {
        let note_emb = vec![1.0_f32, 0.0];
        let topics = make_topic_embeddings(&[("t", 1, vec![1.0, 0.0])]);
        let result = rank_topics_for_note(42, &note_emb, &topics, 0.0, 10);
        assert_eq!(result.len(), 1);
        assert_eq!(result[0].method, LabelingMethod::Embedding);
        assert_eq!(result[0].note_id, 42);
        assert_eq!(result[0].topic_id, 1);
    }
}
